File: output/javascript/tests/39_tagged_instantiation.js

// Test 39: Tagged System Instantiation
// Validates the @@System() syntax for tracked instantiation

export class CalculatorFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class CalculatorFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class CalculatorCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new CalculatorCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class Calculator {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;
    result = 0;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new CalculatorCompartment("Ready");
        this.__next_compartment = null;
        const __frame_event = new CalculatorFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new CalculatorFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new CalculatorFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new CalculatorFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    add(a, b) {
        const __e = new CalculatorFrameEvent("add", {"a": a, "b": b});
        const __ctx = new CalculatorFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_result() {
        const __e = new CalculatorFrameEvent("get_result", null);
        const __ctx = new CalculatorFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_Ready(__e) {
        if (__e._message === "add") {
            const a = __e._parameters?.["a"];
            const b = __e._parameters?.["b"];
            this.result = a + b;
            this._context_stack[this._context_stack.length - 1]._return = this.result;
            return;
        } else if (__e._message === "get_result") {
            this._context_stack[this._context_stack.length - 1]._return = this.result;
            return;
        }
    }
}

export class CounterFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class CounterFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class CounterCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new CounterCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class Counter {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;
    count = 0;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new CounterCompartment("Active");
        this.__next_compartment = null;
        const __frame_event = new CounterFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new CounterFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new CounterFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new CounterFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    increment() {
        const __e = new CounterFrameEvent("increment", null);
        const __ctx = new CounterFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    get_count() {
        const __e = new CounterFrameEvent("get_count", null);
        const __ctx = new CounterFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_Active(__e) {
        if (__e._message === "get_count") {
            this._context_stack[this._context_stack.length - 1]._return = this.count;
            return;
        } else if (__e._message === "increment") {
            this.count = this.count + 1;
        }
    }
}

function main() {
    console.log("=== Test 39: Tagged System Instantiation ===");

    // Tagged instantiation - validated at transpile time
    const calc = new Calculator();
    const counter = new Counter();

    // Test Calculator
    let result = calc.add(3, 4);
    if (result !== 7) {
        console.log(`FAIL: Expected 7, got ${result}`);
        process.exit(1);
    }
    console.log(`Calculator.add(3, 4) = ${result}`);

    result = calc.get_result();
    if (result !== 7) {
        console.log(`FAIL: Expected 7, got ${result}`);
        process.exit(1);
    }
    console.log(`Calculator.get_result() = ${result}`);

    // Test Counter
    counter.increment();
    counter.increment();
    counter.increment();
    const count = counter.get_count();
    if (count !== 3) {
        console.log(`FAIL: Expected 3, got ${count}`);
        process.exit(1);
    }
    console.log(`Counter after 3 increments: ${count}`);

    console.log("PASS: Tagged instantiation works correctly");
}

main();

File: output/csharp/tests/39_tagged_instantiation.cs
using System;
using System.Collections.Generic;


using System;
using System.Collections.Generic;

class CalculatorFrameEvent {
    public string _message;
    public Dictionary<string, object> _parameters;

    public CalculatorFrameEvent(string message) {
        this._message = message;
        this._parameters = null;
    }

    public CalculatorFrameEvent(string message, Dictionary<string, object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class CalculatorFrameContext {
    public CalculatorFrameEvent _event;
    public object _return;
    public Dictionary<string, object> _data;

    public CalculatorFrameContext(CalculatorFrameEvent ev, object defaultReturn) {
        this._event = ev;
        this._return = defaultReturn;
        this._data = new Dictionary<string, object>();
    }
}

class CalculatorCompartment {
    public string state;
    public Dictionary<string, object> state_args;
    public Dictionary<string, object> state_vars;
    public Dictionary<string, object> enter_args;
    public Dictionary<string, object> exit_args;
    public CalculatorFrameEvent forward_event;
    public CalculatorCompartment parent_compartment;

    public CalculatorCompartment(string state) {
        this.state = state;
        this.state_args = new Dictionary<string, object>();
        this.state_vars = new Dictionary<string, object>();
        this.enter_args = new Dictionary<string, object>();
        this.exit_args = new Dictionary<string, object>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    public CalculatorCompartment Copy() {
        CalculatorCompartment c = new CalculatorCompartment(this.state);
        c.state_args = new Dictionary<string, object>(this.state_args);
        c.state_vars = new Dictionary<string, object>(this.state_vars);
        c.enter_args = new Dictionary<string, object>(this.enter_args);
        c.exit_args = new Dictionary<string, object>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class Calculator {
    private List<CalculatorCompartment> _state_stack;
    private CalculatorCompartment __compartment;
    private CalculatorCompartment __next_compartment;
    private List<CalculatorFrameContext> _context_stack;
    public int result = 0;

    public Calculator() {
        _state_stack = new List<CalculatorCompartment>();
        _context_stack = new List<CalculatorFrameContext>();
        __compartment = new CalculatorCompartment("Ready");
        __next_compartment = null;
        CalculatorFrameEvent __frame_event = new CalculatorFrameEvent("$>");
        CalculatorFrameContext __ctx = new CalculatorFrameContext(__frame_event, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    private void __kernel(CalculatorFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            CalculatorFrameEvent exit_event = new CalculatorFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                CalculatorFrameEvent enter_event = new CalculatorFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message == "$>") {
                    __router(forward_event);
                } else {
                    CalculatorFrameEvent enter_event = new CalculatorFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(CalculatorFrameEvent __e) {
        string state_name = __compartment.state;
        if (state_name == "Ready") {
            _state_Ready(__e);
        }
    }

    private void __transition(CalculatorCompartment next) {
        __next_compartment = next;
    }

    public int add(int a, int b) {
        Dictionary<string, object> __params = new Dictionary<string, object>();
        __params["a"] = a;
        __params["b"] = b;
        CalculatorFrameEvent __e = new CalculatorFrameEvent("add", __params);
        CalculatorFrameContext __ctx = new CalculatorFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        var __result = (int) _context_stack[_context_stack.Count - 1]._return;
        _context_stack.RemoveAt(_context_stack.Count - 1);
        return __result;
    }

    public int get_result() {
        CalculatorFrameEvent __e = new CalculatorFrameEvent("get_result");
        CalculatorFrameContext __ctx = new CalculatorFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        var __result = (int) _context_stack[_context_stack.Count - 1]._return;
        _context_stack.RemoveAt(_context_stack.Count - 1);
        return __result;
    }

    private void _state_Ready(CalculatorFrameEvent __e) {
        if (__e._message == "add") {
            var a = (int) __e._parameters["a"];
            var b = (int) __e._parameters["b"];
            this.result = a + b;
            _context_stack[_context_stack.Count - 1]._return = this.result;
            return;
        } else if (__e._message == "get_result") {
            _context_stack[_context_stack.Count - 1]._return = this.result;
            return;
        }
    }
}

class CounterFrameEvent {
    public string _message;
    public Dictionary<string, object> _parameters;

    public CounterFrameEvent(string message) {
        this._message = message;
        this._parameters = null;
    }

    public CounterFrameEvent(string message, Dictionary<string, object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class CounterFrameContext {
    public CounterFrameEvent _event;
    public object _return;
    public Dictionary<string, object> _data;

    public CounterFrameContext(CounterFrameEvent ev, object defaultReturn) {
        this._event = ev;
        this._return = defaultReturn;
        this._data = new Dictionary<string, object>();
    }
}

class CounterCompartment {
    public string state;
    public Dictionary<string, object> state_args;
    public Dictionary<string, object> state_vars;
    public Dictionary<string, object> enter_args;
    public Dictionary<string, object> exit_args;
    public CounterFrameEvent forward_event;
    public CounterCompartment parent_compartment;

    public CounterCompartment(string state) {
        this.state = state;
        this.state_args = new Dictionary<string, object>();
        this.state_vars = new Dictionary<string, object>();
        this.enter_args = new Dictionary<string, object>();
        this.exit_args = new Dictionary<string, object>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    public CounterCompartment Copy() {
        CounterCompartment c = new CounterCompartment(this.state);
        c.state_args = new Dictionary<string, object>(this.state_args);
        c.state_vars = new Dictionary<string, object>(this.state_vars);
        c.enter_args = new Dictionary<string, object>(this.enter_args);
        c.exit_args = new Dictionary<string, object>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class Counter {
    private List<CounterCompartment> _state_stack;
    private CounterCompartment __compartment;
    private CounterCompartment __next_compartment;
    private List<CounterFrameContext> _context_stack;
    public int count = 0;

    public Counter() {
        _state_stack = new List<CounterCompartment>();
        _context_stack = new List<CounterFrameContext>();
        __compartment = new CounterCompartment("Active");
        __next_compartment = null;
        CounterFrameEvent __frame_event = new CounterFrameEvent("$>");
        CounterFrameContext __ctx = new CounterFrameContext(__frame_event, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    private void __kernel(CounterFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            CounterFrameEvent exit_event = new CounterFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                CounterFrameEvent enter_event = new CounterFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message == "$>") {
                    __router(forward_event);
                } else {
                    CounterFrameEvent enter_event = new CounterFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(CounterFrameEvent __e) {
        string state_name = __compartment.state;
        if (state_name == "Active") {
            _state_Active(__e);
        }
    }

    private void __transition(CounterCompartment next) {
        __next_compartment = next;
    }

    public void increment() {
        CounterFrameEvent __e = new CounterFrameEvent("increment");
        CounterFrameContext __ctx = new CounterFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        _context_stack.RemoveAt(_context_stack.Count - 1);
    }

    public int get_count() {
        CounterFrameEvent __e = new CounterFrameEvent("get_count");
        CounterFrameContext __ctx = new CounterFrameContext(__e, null);
        _context_stack.Add(__ctx);
        __kernel(_context_stack[_context_stack.Count - 1]._event);
        var __result = (int) _context_stack[_context_stack.Count - 1]._return;
        _context_stack.RemoveAt(_context_stack.Count - 1);
        return __result;
    }

    private void _state_Active(CounterFrameEvent __e) {
        if (__e._message == "get_count") {
            _context_stack[_context_stack.Count - 1]._return = this.count;
            return;
        } else if (__e._message == "increment") {
            this.count = this.count + 1;
        }
    }
}

class Program {
    static void Main(string[] args) {
        Console.WriteLine("=== Test 39: Tagged System Instantiation ===");

        // Tagged instantiation - validated at transpile time
        var calc = new Calculator();
        var counter = new Counter();

        // Test Calculator
        var result = (int)calc.add(3, 4);
        if (result != 7) {
            Console.WriteLine("FAIL: Expected 7, got " + result);
            Environment.Exit(1);
        }
        Console.WriteLine("Calculator.add(3, 4) = " + result);

        result = (int)calc.get_result();
        if (result != 7) {
            Console.WriteLine("FAIL: Expected 7, got " + result);
            Environment.Exit(1);
        }
        Console.WriteLine("Calculator.get_result() = " + result);

        // Test Counter
        counter.increment();
        counter.increment();
        counter.increment();
        var count = (int)counter.get_count();
        if (count != 3) {
            Console.WriteLine("FAIL: Expected 3, got " + count);
            Environment.Exit(1);
        }
        Console.WriteLine("Counter after 3 increments: " + count);

        Console.WriteLine("PASS: Tagged instantiation works correctly");
    }
}
